ui/src/model/geometry.ts

```diff
@@ -31,7 +31,63 @@ export type BoundingBox = {
     y: Range<number>;
 };
 
-export type Rectangle = Polygon;
+export type Rectangle = OlPolygon;
+
+export type Polygon = {
+    points: Point[];
+    bouundingBox?: BoundingBox;
+};
+
+export const coordsToPolygon = (coords: Coordinate[]): Polygon => ({
+    points: coords.map(([x, y]) => ({ x: x ?? 0, y: y ?? 0 })),
+});
+
+const shortenLineFromBothEnds = (line: Line, amount: number): Line => {
+    const hypot = Math.hypot(line.end.x - line.start.x, line.end.y - line.start.y);
+    const normalized =
+        hypot === 0
+            ? { x: 0, y: 0 }
+            : {
+                  x: (line.end.x - line.start.x) / hypot,
+                  y: (line.end.y - line.start.y) / hypot,
+              };
+    const adjustedStart = {
+        x: line.start.x + normalized.x * amount,
+        y: line.start.y + normalized.y * amount,
+    };
+    const adjustedEnd = {
+        x: line.end.x - normalized.x * amount,
+        y: line.end.y - normalized.y * amount,
+    };
+    return { start: adjustedStart, end: adjustedEnd };
+};
+
+export function isValidPolygon(coords: Coordinate[], ignoreLastSegment: boolean) {
+    const allPolygonSegments = coords
+        .slice(0, -1)
+        .map((startCoordinate, index, coordinatesExceptLoopback) => {
+            const endCoordinate = expectDefined(
+                coordinatesExceptLoopback[(index + 1) % coordinatesExceptLoopback.length],
+            );
+            return [startCoordinate, endCoordinate] as const;
+        });
+    const relevantSegments = ignoreLastSegment
+        ? allPolygonSegments.slice(0, -1)
+        : allPolygonSegments;
+
+    return !relevantSegments
+        .map(([start, end]) =>
+            shortenLineFromBothEnds(
+                createLine(coordsToPoint(start), coordsToPoint(end)),
+                0.0000001,
+            ),
```

Comment:
Tässä kanssa eräässä dailyssä puhuttu dirty hack jonka tiketöin omaksi jutukseen: Liukuluvut hajottavat polygonin reunojen päällekkäisyyden tutkimisen algoritmin (polygonin viereisten viivojen päät alkavat osua toisiinsa pyöristysvirheiden takia). Tämän kiertämiseen riittää todella lyhyt typistys polygonin loppupäiden viivoille, joten lyhennetään viivoja 0.0000001 koordinaattiyksiköllä (eli TM35FIN:issä noin 0.1 mikrometrillä)